src/EnergyPlus/PlantLoopHeatPumpEIR.cc

```diff
@@ -1289,4 +1290,1373 @@ void EIRPlantLoopHeatPump::oneTimeInit(EnergyPlusData &state)
+{
+    if (state.dataEIRFuelFiredHeatPump->getInputsFFHP) {
+        EIRFuelFiredHeatPump::processInputForEIRPLHP(state);
+        EIRFuelFiredHeatPump::pairUpCompanionCoils(state);
+        state.dataEIRFuelFiredHeatPump->getInputsFFHP = false;
+    }
+
+    for (auto &plhp : state.dataEIRFuelFiredHeatPump->heatPumps) {
+        if (plhp.name == UtilityRoutines::MakeUPPERCase(hp_name) && plhp.EIRHPType == hp_type_of_num) {
+            return &plhp;
+        }
+    }
+
+    ShowFatalError(state, "EIR Fuel-Fired Heat Pump factory: Error getting inputs for PLFFHP named: " + hp_name);
+    return nullptr; // LCOV_EXCL_LINE
+}
+
+void EIRFuelFiredHeatPump::pairUpCompanionCoils(EnergyPlusData &state)
+{
+    for (auto &thisHP : state.dataEIRFuelFiredHeatPump->heatPumps) {
+        if (!thisHP.companionCoilName.empty()) {
+            auto thisCoilName = UtilityRoutines::MakeUPPERCase(thisHP.name);
+            auto &thisCoilType = thisHP.EIRHPType;
+            auto targetCompanionName = UtilityRoutines::MakeUPPERCase(thisHP.companionCoilName);
+            for (auto &potentialCompanionCoil : state.dataEIRFuelFiredHeatPump->heatPumps) {
+                auto &potentialCompanionType = potentialCompanionCoil.EIRHPType;
+                auto potentialCompanionName = UtilityRoutines::MakeUPPERCase(potentialCompanionCoil.name);
+                if (potentialCompanionName == thisCoilName) {
+                    // skip the current coil
+                    continue;
+                }
+                if (potentialCompanionName == targetCompanionName) {
+                    if (thisCoilType == potentialCompanionType) {
+                        ShowSevereError(state,
+                                        "Invalid companion specification for EIR Plant Loop Fuel-Fired Heat Pump named \"" + thisCoilName + "\"");
+                        ShowContinueError(state, "For heating objects, the companion must be a cooling object, and vice-versa");
+                        ShowFatalError(state, "Invalid companion object causes program termination");
+                    }
+                    thisHP.companionHeatPumpCoil = &potentialCompanionCoil;
+                    break;
+                }
+            }
+            if (!thisHP.companionHeatPumpCoil) {
+                ShowSevereError(state, "Could not find matching companion heat pump coil.");
+                ShowContinueError(state, "Base coil: " + thisCoilName);
+                ShowContinueError(state, "Looking for companion coil named: " + targetCompanionName);
+                ShowFatalError(state, "Simulation aborts due to previous severe error");
+            }
+        }
+    }
+}
+
+void EIRFuelFiredHeatPump::processInputForEIRPLHP(EnergyPlusData &state)
+{
+    struct ClassType
+    {
+        DataPlant::PlantEquipmentType thisType;
+        std::string nodesType;
```

Comment:
A type should not be a `string`.

Reply:
ping @jcyuan2020

Reply:
Actually this field is actually used in `BanchNodeConnections::TestCompSet()` as the string for warning message (the 6th parameter `std::string &Description`: 
```
void TestCompSet(EnergyPlusData &state,
                 std::string_view const CompType, // Component Type
                 std::string_view CompName,       // Component Name
                 std::string const &InletNode,    // Inlet Node Name
                 std::string const &OutletNode,   // Outlet Node Name
                 std::string const &Description   // Description of Node Pair (for warning message)
)
```
Since this is the only place that this field is used, and only used for warning messages, I am not sure if it is worthwhile to officially do an `enum` for it and then do the conversion trick to get the corresponding string name---actually there is no need for a standard (enum or string) name here either.